packages/tasit-identity-contract/src/GnosisSafe.test.js

```diff
@@ -60,138 +61,211 @@ describe("GnosisSafe", () => {
+
+        const errorListener = async message => {
+          onError();
+          action.unsubscribe();
+        };
+
+        // Note: Some error events are been trigger only from the confirmationListener
+        // See more: https://github.com/tasitlabs/TasitSDK/issues/253
+        const confirmationListener = () => {};
+
+        action.on("error", errorListener);
+        action.on("confirmation", confirmationListener);
+
+        await action.waitForNonceToUpdate();
+
+        await mineBlocks(provider, 1);
+
+        expect(onError.callCount).to.equal(1);
+        confirmBalances(erc20, [GNOSIS_SAFE_ADDRESS], [balanceBefore]);
+      });
+
+      describe("test cases that need ERC20 spending approval for ephemeral account", () => {
+        // TODO: Move to tasit-link-wallet
+        beforeEach(
+          "contract-based accounts' owner should approve an ephemeral account to spend funds",
+          async () => {
+            const tokenAddress = ERC20_ADDRESS;
+            const { address: spenderAddress } = ephemeralAccount;
+
+            gnosisSafe.setSigners([gnosisSafeOwner]);
+            gnosisSafe.setWallet(gnosisSafeOwner);
+            const action = gnosisSafe.approveERC20(
+              tokenAddress,
+              spenderAddress,
+              SMALL_AMOUNT
+            );
+            await action.waitForNonceToUpdate();
+
+            await mineBlocks(provider, 1);
+
+            const amountAllowed = await erc20.allowance(
+              GNOSIS_SAFE_ADDRESS,
+              spenderAddress
+            );
+
+            expect(`${amountAllowed}`).to.equal(`${SMALL_AMOUNT}`);
+          }
+        );
+
+        it("ephemeral account should transfer allowed funds from the contract-based account", async () => {
+          const balanceBefore = await erc20.balanceOf(GNOSIS_SAFE_ADDRESS);
+          const { address: toAddress } = someone;
+
+          erc20.setWallet(ephemeralAccount);
+          const action = erc20.transferFrom(
+            GNOSIS_SAFE_ADDRESS,
+            toAddress,
+            SMALL_AMOUNT
+          );
+
+          await action.waitForNonceToUpdate();
+
+          await mineBlocks(provider, 1);
+
+          const expectedBalance = balanceBefore.sub(SMALL_AMOUNT);
+          confirmBalances(erc20, [GNOSIS_SAFE_ADDRESS], [expectedBalance]);
+          confirmBalances(erc20, [toAddress], [SMALL_AMOUNT]);
+        });
+      });
+    });
+  });
 
-      // Note: Some error events are been trigger only from the confirmationListener
-      // See more: https://github.com/tasitlabs/TasitSDK/issues/253
-      const confirmationListener = () => {};
+  describe("test cases that need NFT deposit to the contract-based account", async () => {
+    const tokenId = 1;
 
-      execTxAction.on("error", errorListener);
-      execTxAction.on("confirmation", confirmationListener);
+    beforeEach("faucet", async () => {
+      await erc721Faucet(nft, root, GNOSIS_SAFE_ADDRESS, tokenId);
```

Comment:
Reading `root` here I forgot what we meant by it for a second - maybe a more explicit name is warranted

Reply:
Okay, I'll use `minter`, okay?

Reply:
Sounds good!

Reply:
Done.

Reply:
Thanks! Resolving